HW1/P5/P5_bfs.py

```diff
@@ -0,0 +1,221 @@
+import pickle
+import string
+import re
+import urllib2
+import itertools
+import math as ma
+import os
+import pyspark
+
+sc = pyspark.SparkContext()
+sc.setLogLevel('WARN')
+
+def link_string_to_KV(s):
+    src, dests = s.split(': ')
+    dests = [int(to) for to in dests.split(' ')]
+    return (int(src), dests)
+
+
+def fun(i):
+    return lambda x: (x[0], i + 1)
+
+
+def copartitioned(RDD1, RDD2):
+    "check if two RDDs are copartitioned"
+    return RDD1.partitioner == RDD2.partitioner
+
+
+def find_distance(node, dist):
+        for i in dist:
+                if i[0] == node:
+                        position = i[1]
+        return position
+
+
+links = sc.textFile('s3://Harvard-CS205/wikipedia/links-simple-sorted.txt', 32)
+page_names = sc.textFile('s3://Harvard-CS205/wikipedia/titles-sorted.txt', 32)
+
+page_names = page_names.zipWithIndex().map(lambda (n, id): (n, id + 1))
+
+Kevin_Bacon = page_names.filter(lambda (K, V): K == 'Kevin_Bacon').collect()
+assert len(Kevin_Bacon) == 1
+Kevin_Bacon = unicode(Kevin_Bacon[0][1])
+
+Harvard_University = page_names.filter(lambda (K, V): K == 'Harvard_University').collect()
+assert len(Harvard_University) == 1
+Harvard_University = unicode(Harvard_University[0][1])
+
+regex = re.compile(r'[%s\s]+' % re.escape(string.punctuation))
+split_links = links.map(lambda s: regex.split(s))
+
+full_rdd2 = split_links.map(lambda x: (x[0], x[1:])).partitionBy(256).cache()
+
+
+############################################# Kevin to Harvard #############################
+
+start_node = Kevin_Bacon
+end_node = Harvard_University
+
+start = [(start_node, 0)] 
+rdd = sc.parallelize(start).partitionBy(256)
+i = 0
+
+num_runs = 4
+
+while i < num_runs:
+    assert copartitioned(full_rdd2, rdd)
+    
+    rdd_parent = rdd.filter(lambda x: x[1] == i).join(full_rdd2).mapValues(lambda x: x[1])
+    assert copartitioned(rdd, rdd_parent)
+    
+    rdd_children = rdd_parent.flatMap(lambda x: x[1]).map(lambda x: (x, 1)).subtractByKey(rdd).distinct().partitionBy(256)

```

Comment:
Similar to above comments - this code makes sense, but it took me a little while to understand what your purpose was here. It would have been much cleaner if you had a clear comment saying what this step does, or breaking it up into several steps as the teaching staff suggested.